pcsx2/USB/USB.cpp

```diff
@@ -20,286 +20,174 @@
-s64 remaining = 0;
+static OHCIState* s_qemu_ohci = nullptr;
+static USBDevice* s_usb_device[USB::NUM_PORTS] = {};
+static const DeviceProxy* s_usb_device_proxy[USB::NUM_PORTS] = {};
 
-#if defined(_WIN32)
-HWND gsWnd = nullptr;
-#elif defined(__linux__)
-#include "gtk.h"
-#include <gdk/gdkx.h>
-#include <X11/X.h>
-Display* g_GSdsp = nullptr;
-Window g_GSwin;
-#endif
+static s64 s_usb_clocks = 0;
+static s64 s_usb_remaining = 0;
 
-Config::Config()
-	: Log(0)
-{
-}
+int64_t g_usb_frame_time = 0;
+int64_t g_usb_bit_time = 0;
+int64_t g_usb_last_cycle = 0;
 
-//Simpler to reset and reattach after USBclose/USBopen
-void Reset()
+std::string USBGetConfigSection(int port)
 {
-	if (qemu_ohci)
-		ohci_hard_reset(qemu_ohci);
+	return fmt::format("USB{}", port + 1);
 }
 
-void OpenDevice(int port)
+OHCIPort& USB::GetOHCIPort(u32 port)
 {
-	//TODO Pass pDsp to open probably so dinput can bind to this HWND
-	if (usb_device[port] && usb_device[port]->klass.open)
-		usb_device[port]->klass.open(usb_device[port] /*, pDsp*/);
+	// Apparently the ports on the hub are swapped.
+	// Get this wrong and games like GT4 won't spin your wheelz.
+	const u32 rhport = (port == 0) ? 1 : 0;
+	return s_qemu_ohci->rhport[rhport];
 }
 
-static void CloseDevice(int port)
+bool USB::CreateDevice(u32 port)
 {
-	if (usb_device[port] && usb_device[port]->klass.close)
-		usb_device[port]->klass.close(usb_device[port]);
-}
+	const Pcsx2Config::USBOptions::Port& portcfg = EmuConfig.USB.Ports[port];
+	const DeviceProxy* proxy = RegisterDevice::instance().Device(portcfg.DeviceType);
+	if (!proxy)
+		return true;
 
-void DestroyDevice(int port)
-{
-	if (qemu_ohci && qemu_ohci->rhport[port].port.dev)
+	DevCon.WriteLn("(USB) Creating a %s in port %u", proxy->Name(), port + 1);
+	USBDevice* dev;
 	{
-		qemu_ohci->rhport[port].port.dev->klass.unrealize(qemu_ohci->rhport[port].port.dev);
-		qemu_ohci->rhport[port].port.dev = nullptr;
+		auto lock = Host::GetSettingsLock();
+		dev = proxy->CreateDevice(*Host::GetSettingsInterface(), port, portcfg.DeviceSubtype);
 	}
-	else if (usb_device[port])
-		usb_device[port]->klass.unrealize(usb_device[port]);
-
-	usb_device[port] = nullptr;
-}
-
-void DestroyDevices()
-{
-	for (int i = 0; i < 2; i++)
+	if (!dev)
 	{
-		CloseDevice(i);
-		DestroyDevice(i);
+		Console.Error("Failed to create USB device in port %u (%s)", port + 1, proxy->Name());
+		return false;
 	}
-}
-
-static USBDevice* CreateDevice(DeviceType index, int port)
-{
-	USBDevice* device = nullptr;
 
-	if (index == DEVTYPE_NONE)
-		return nullptr;
-
-	DeviceProxyBase* devProxy = RegisterDevice::instance().Device(index);
-	if (devProxy)
-		device = devProxy->CreateDevice(port);
-	else
-		Console.WriteLn(Color_Red, "Device %d: Unknown device type", 1 - port);
-
-	if (!device)
+	if (dev->klass.open && dev->klass.open(dev) != 0)
 	{
+		Console.Error("Failed to open USB device in port %u (%s)", port + 1, proxy->Name());
+		if (dev->klass.unrealize)
+			s_usb_device[port]->klass.unrealize(dev);
```

Comment:
100% of errors (2 reports) are caused by this line. I presume you meant dev->klass.unrealize(dev);? s_usb_device is null for both ports and isn't set until after this.